repair/repair.cc

```diff
@@ -669,28 +673,28 @@ future<> repair::shard_repair_task_impl::repair_range(const dht::token_range& ra
     if (neighbors.empty()) {
         auto status = "skipped_no_followers";
         rlogger.warn("repair[{}]: Repair {} out of {} ranges,  keyspace={}, table={}, range={}, peers={}, live_peers={}, status={}",
-                global_repair_id.uuid(), ranges_index, ranges_size(), _status.keyspace, table_names(), range, neighbors, live_neighbors, status);
+                global_repair_id.uuid(), ranges_index, all_ranges, _status.keyspace, table.name, range, neighbors, live_neighbors, status);
         co_return;
     }
     rlogger.debug("repair[{}]: Repair {} out of {} ranges, keyspace={}, table={}, range={}, peers={}, live_peers={}",
-        global_repair_id.uuid(), ranges_index, ranges_size(), _status.keyspace, table_names(), range, neighbors, live_neighbors);
-    co_await mm.sync_schema(db.local(), neighbors);
+        global_repair_id.uuid(), ranges_index, all_ranges, _status.keyspace, table.name, range, neighbors, live_neighbors);
+    co_await _local_repair_data.mm.sync_schema(_local_repair_data.db.local(), neighbors);
     sstring cf;
     try {
-        cf = db.local().find_column_family(table_id).schema()->cf_name();
+        cf = _local_repair_data.db.local().find_column_family(table.id).schema()->cf_name();
     } catch (replica::no_such_column_family&) {
         co_return;
     }
     // Row level repair
-    if (dropped_tables.contains(cf)) {
+    if (_local_repair_data.dropped_tables.contains(cf)) {
         co_return;
     }
     try {
-        co_await repair_cf_range_row_level(*this, cf, table_id, range, neighbors);
+        co_await repair_cf_range_row_level(_local_repair_data, cf, table.id, range, neighbors);
     } catch (replica::no_such_column_family&) {
-        dropped_tables.insert(cf);
+        _local_repair_data.dropped_tables.insert(cf);
```

Comment:
Looking at all those reference to the shard repair task inside the new range repair task. It makes me to ask do we really need the range task in the first place. Yes, it is fancy have the task model everywhere. But is it enough to end at the shard repair task level?

Reply:
Table tasks allow to have more precise insight in repair process as we can see the progress of a particular table repair. So it actually depends on whether you find table progress useful for users or not.

Yes, the references are nasty, but `_local_repair_data`'s fields contain data relevant to the whole repair and not just one table. Thinking of ways to make it more bearable, following solutions come to my mind:
- Keep references to shard task field and not to shard task itself - I don't really like this one, quite a mess and unnecessary waste of memory. 
- Enclose the shard-specific data in a nested class and keep a shared_ptr to it in both shard and table tasks. - This solves the "is it still alive?" problem, but we still have the references with clearer meaning though. This way we would go back to "many different classes with similar intention" which we tried to avoid by merging repair_info with shard task.
- Drop the commit that moves `repair_range` to table task from shard task. It is rather not table-specific and we could treat as a for-all-ranges-regardless-the-table machinery.

I like the last one the most. What do you think?

Reply:
The percentage of ranges we have repaired is useful. Which specific range is finished or not is not that important. Also, we have two levels inside shard task (tables and ranges). Should we also introduce yet another table_repair_task_impl along with the range_repair_task_impl under shard_repair_task_impl? I think we should not.  Further, we have multiple sub ranges get repaired inside each range (range_repair_task_impl), should we create subrange_repair_task_impl to track them? I think no. 

I wish we could implement all the percentage api for the task module inside the repair::shard_repair_task_impl and keep the hierarchy simple.  I hope this is what you meant in item 3 above.

Reply:
Oh, actually I didn't mean to create range_repair_task_impl (even though I did at first) and would stay with only per table tasks under shard ones. But ok, if you prefer so, I will get rid of table tasks and stop on shard level

Reply:
I'm not sure it's a good idea to hide the per-table task.
I think it's interesting to the user / field engineering if repair is slow or hung on a particular table.

Reply:
Without the table_repair_task_impl, it is a simple for-loop to work on the tables we want to repair. With table_repair_task_impl we need to create all the task_impl objects. I think it is overkill to create all such task machinery for debugging which table or range we are working on.  We can set a flag like _current_table and _current_range to store such info. Perhaps we can extend the task machinery to allow exposing a key value  per task.

For example per share_repair_task_impl, 

"current_reapir_table: "mytable"
"current_reapir_range": "100:200"

This is much more flexible  than looking at which pending tasks to figure out this info. Imagine there are 10K ranges to repair in total for 2 tables.  No one is going to look at individual range_repair_task_impl. 

@Deexie ^

Reply:
> Oh, actually I didn't mean to create range_repair_task_impl (even though I did at first) and would stay with only per table tasks under shard ones. But ok, if you prefer so, I will get rid of table tasks and stop on shard level

Great!  Btw, in practice, the management tool would send one table per repair job in most cases.

Reply:
> No one is going to look at individual range_repair_task_impl.

This is the reason low-level task types are considered as "internal" and they won't show up to the user without a flag to ask for them.  But id they are not required and we can do without them, that's good too.

In the tablets world, a tablet is a sub-range of the token ring, and it would make sense to have a task per tablet (it's also on one shard so its lighter on the system).  If and when tablet repair will be done autonomously and periodically (especially with incremental repair) we'd probably have a top-level task on the "repair coordinator" node when a periodic repair starts.  I think we won't need tasks on the replicas since they would probably just react to the repair protocol commands sent to them, and keep some local state (like `repair_meta` today).